### ISSUE TITLE
Adding a custom routing convention

### ISSUE THREAD

**Issue by arch-imp** [Microsoft/aspnet-api-versioning#298]
I'd like to add a custom routing convention.  An implementation of IODataRoutingConvention.  I'm currently using the overload of MapVersionedODataRoutes that accepts a configureAction parameter.  Where would I plug in my IODataRoutingConvention implementation?

**Comment by commonsensesoftware** [comment]
Unfortunately, OData is somewhat limited in how you're able to set that up. I believe you only have two feasible options:

1. Use one of the methods that accepts the conventions.
2. Completely replace all conventions in the container. If you choose this option, you'll break API versioning unless you replicate it's convention setup.

I'm not sure your exact scenario. If you take option 2, you'll want to look at [HttpConfigurationExtensions.cs](../blob/master/src/Microsoft.AspNet.OData.Versioning/System.Web.Http/HttpConfigurationExtensions.cs). Most of the complexity in the library involves building the routes.  I'm open to suggestions on how this can be more flexible. The current design is largely parallel to the built-in OData design.

**Comment by commonsensesoftware** [comment]
I see what you're doing. No one has mentioned this behavior, but it would seem there is a limitation or, at least difficulty, in composing custom conventions. It sounds like it's worth adding a new method overload to address it.

By chance, have you tried using attribute routing? I suspect you can make this work with attribute routing and then you wouldn't need a custom route convention. That doesn't negate the need to have more flexibility in route configurations, but if it works, it could immediately unblock you without having to do a bunch of work.

**Comment by commonsensesoftware** [comment]
Theoretically, that _should_ be possible. API versioning always adds conventions because it has its own custom conventions to make versioning work in OData. Regardless, I think you've demonstrated that this is something that is way more difficult to setup than it should be. I've queue up #308 to track this and add it to the next release.

**Comment by arch-imp** [comment]
I couldn't find a nice way to do this.  At least, one that I understood without duplicating a lot of library code.  I'm continuing to prototype ideas.

If I were you, I'd work through the OData samples at https://github.com/OData/ODataSamples/tree/master/WebApi/v4 and add versioning to each of them.  The result would be a very robust versioning library. 

In my instance, the sample I would like to see versioned is https://github.com/OData/ODataSamples/tree/master/WebApi/v4/ODataCompositeKeySample.  If that sample can have the versionin library applied to it, then my requirement could be met.

**Comment by commonsensesoftware** [comment]
Fair. As a long-time OData user and advocate of the protocol, I have a history with both the Microsoft stack and the various people that have been involved with it over the years. It's really disappointing the shape it's been left in. The entire stack has seen numerous team owners. That said, the documentation, examples, etc are often not current.  The example you are referring to is clearly one of them.

Composite keys have been supported for quite some time using attributes (ex: `[ODataRoute]`). This was the only supported way for a while. Attributes also tend to be preferrable as they are more explicit and descriptive. OData is a little unique in that to even use the route attributes, you still have to know something about how the conventions work. This results in a number OData service authors wanting to use the standard OData conventions and have it _just work_. This gap beween attributes and implicit conventions was addressed directly in the platform starting in version 6.0. You can see the official documentation [here](http://odata.github.io/WebApi/#13-06-KeyValueBinding). This capability is intrinsically supported by API versioning and the API explorer support without any additional custom, user-defined conventions.

Despite all that, there _could_ still be a need to have a custom convention or to otherwise change the default conventions. Today, the API versioning design doesn't make it very easy to change or modify them. There was no specific reason for that, but there is definitely some risk in changing the conventions. If it's not done carefully, API versioning will break because it needs to add its own custom conventions to make OData do the right thing. Regardless, convention modification should be open, not closed. Changing it correctly is the onus of the developer.

**Comment by commonsensesoftware** [comment]
Sounds like you are your way. I opened up feature #308, which will be included in forthcoming v3.0. The beta will be ready soon. If you're curious what it will look like, you can checkout the updates in the v3.0 branch [here](../blob/v3/src/Microsoft.AspNet.OData.Versioning/System.Web.Http/HttpConfigurationExtensions.cs#L114). This will build the API versioning OData conventions, but still give devs a hook to modify them before they are placed in the container.

If you have additional questions, feel free to re-open this or open an new issue. Thanks.

**Comment by arch-imp** [comment]
I've had to drop this part of our API due to other priorities.  I'm anxious to get back to it.  I'll look at the new version hopefully soon.  Thanks for the quick turn around!
